---
title: "Estimating mutation rate for SARS-CoV-2"
author: "Azim Ansari"
date: "03/11/2021"
output: 
  html_document:
    toc: true
---

# Problem

We would like to estimate the mutation rate of SARS-CoV-2 (COVID virus). To do this we will use the
original Wuhan sequence that was collected in December of 2019 and a sequence from September of 2021 in the UK.
In this example we will calculate the mutation rate as the number of mutations per genome per month ($\frac{\text{mutations}}{\text{genome}*\text{month}}$), but you can estimate the mutation rate as the number of mutations per nucleotide site per day or some other unit.

# Point estimate

A simple way of estimating the mutation rate is to:

1. Count the number of nucleotide differences between the original Wuhan sample and the sample collected from the UK in Sep 2021.
2. Divide this number by the length of time between the two samples (decide on unit of time, could be in years, months, weeks, etc.)

let's do the above and see what we get.

### Calculate mutation rate

The SARS-CoV-2 sequences are stored in a fasta file. If you don't know about fasta format you can read about them [here](https://zhanggroup.org/FASTA/).

To read nucleotide sequences into `R` you need to install a package called `seqinr`. The whole genome sequences are already aligned. The first sequence is the original Wuhan sequence which was collected in December 2019 and the other 3 sequences were collected in September 2021 in Egypt, UK and India.

```{r }
# install.packages('seqinr')
library('seqinr')

# read the sequences
sars_seqs = read.fasta('./covid_3_sequences_from_sep_2021_and_original_wuhan_sequence.fasta')
summary(sars_seqs) # View a summary of the sequence data.

# calculate the number of nucleotide differences between Wuhan and each of the new sequences.
num_of_mutations_egypt = sum(sars_seqs$`NC_045512|china|2019-12` != sars_seqs$`OK104629|Egypt|2021-09-03`)
num_of_mutations_UK = sum(sars_seqs$`NC_045512|china|2019-12` != sars_seqs$`OU732380|UK|2021-09-11`)
num_of_mutations_India = sum(sars_seqs$`NC_045512|china|2019-12` != sars_seqs$`OK356443|India|2021-09-25`)

# Print the number of mutations per month for each of the new sequences.
number_of_months = 21
num_of_mutations_egypt / number_of_months
num_of_mutations_UK / number_of_months
num_of_mutations_India / number_of_months 
```

For each sequence we get a slightly different point estimate. We have no idea about the uncertainty associated with this point estimate. Given the data how likely is it that the mutation rate is less than 3 mutations per genome per months?

To answer the above question we need to have a statistical model. Let's do that.

# Model for mutation rate

First of all we should think of what kind of probability model would fit our data best. A commonly used model that fits the observed number of mutations is the Poisson distribution.
Poisson distribution has the following PMF with the parameter $\lambda t$. 

\[
\text{P}(X = k) = \frac{e^{\lambda t} ({\lambda t})^k}{k!}
\]

Where $\lambda$ is the mutation rate and $t$ is the length of time between the two sequences. So if you know the mutation rate, then you can calculate the probability of observing a specific number of mutations after certain length of time using the above formula.

### Posterior inference

We are assuming that the distribution of the mutation numbers given that we know the mutation rate is Poisson.

\[
X = \text{number of mutations}\\
\lambda = \text{mutation rate} ~~(\frac{\text{mutations}}{\text{genome} * \text{month}})\\
X|\lambda \sim \text{Pois}(\lambda t)
\]

Here we know what $X$ (the number of mutations) is and would like to infer the distribution of $\lambda$. Bayes theorem is how we can estimate the distribution of $\lambda$.

\[
f(\lambda|X=k) =\frac{P(X=k|\lambda) f(\lambda)}{P(X = k)}
\]

We have already decided on what is the data generating process i.e. the likelihood i.e. what model fits our data best. We next need to decide on the prior distribution of $\lambda$ or $f(\lambda)$. And with the likelihood and the prior we can calculate the posterior. Before we do that let's plot the likelihood.

### Likelihood

What is the likelihood? It is simply the probability of observing the data for different parameter values. In this instance for different mutation rates what is the probability of observing 42 mutations (number of nucleotide differences between Wuhan sequence and the Egyptian sequence).
What does the likelihood looks like when we use different sequences?

```{r}
lambda_range = seq(0,6,by=0.001)

likelihood_egypt = dpois(num_of_mutations_egypt,lambda_range * number_of_months)
plot(lambda_range,likelihood_egypt,xlab = 'mutation rate (per genome per month)',ylab = 'likelihood',cex = 0.5,ylim = c(0,0.07))

likelihood_india = dpois(num_of_mutations_India,lambda_range * number_of_months)
points(lambda_range,likelihood_india,xlab = 'mutation rate (per genome per month)',ylab = 'likelihood',col='darkgray',cex = 0.5)

likelihood_uk = dpois(num_of_mutations_UK,lambda_range * number_of_months)
points(lambda_range,likelihood_uk,xlab = 'mutation rate (per genome per month)',ylab = 'likelihood',col='lightgray',cex = 0.5)

legend(x=5,y=0.05,legend=c('Egypt','India','UK'),col =c('black','darkgray','lightgray'),pch=19)

```

### Prior and prior predictive checks

let's assume the prior distribution on $\lambda$ is uniform on 0 to 50. That means we think that the number of mutations per genome per month could be anything between 0 and 50 with equal probability. Does this make sense? How can we check to see if this makes sense?

let's do prior predictive checks to see what kind of data we are likely to see given the prior distribution for $\lambda$. We can generate random samples from the prior distribution for mutation rate and use those samples to generate the observed number of mutations. We can then check to see if those numbers make sense based on our knowledge of the world.

```{r}
# sample from the prior.
prior_samples_lambda = runif(1000,0,50)
t = number_of_months

prior_num_mutations = rpois(1000,prior_samples_lambda*t)
hist(prior_num_mutations,xlab = 'observed number of mutations in 21 months')
```

Does the above make sense to you? Your prior says that you are as likely to have few mutations as having nearly 1000 mutations between the original sequence and the sequence from Sep 2021.

We can look at the mutation rates for other known RNA viruses. For instance the mutation rate for Hepatitis C virus (HCV) is around 1 mutation per genome per month. It is important to remember that HCV genome is around 10 Kb long while SARS-CoV-2 genome is around 30 Kb long. So we would expect around 3 mutations per genome per months and a total of around 63 mutations in 21 months.

As a prior distribution for the mutation rate we could use Gamma distribution. There is a special relationship between Gamma and Poisson distributions called `conjugacy`: If we have a Gamma prior distribution on $\lambda$ and data are conditionally Poisson given $\lambda$  then going from prior to posterior we don't leave the family of the Gamma distributions and we say that Gamma is the conjugate prior of the Poisson. In other words if we use a Gamma distribution as the prior for $\lambda$ then the the posterior for $\lambda$ is also going to be a Gamma distribution. The Gamma distribution has two parameters $\lambda \sim \text{Gamma}(r_0,b_0)$. The parameters $r_0$ and $b_0$ in the prior can be interpreted as prior $r_0$ mutation in prior $b_0$ months. What kind of parameter values make sense to you? Maybe around 3 mutation per genome per month using HCV as a guide. $\lambda \sim \text{Gamma}(3,1)$

\[
X = \text{number of mutations}\\
\lambda = \text{mutation rate} ~~(\frac{\text{mutations}}{\text{genome} * \text{month}})\\
\lambda \sim \text{Gamma}(r_0,b_0)\\
X|\lambda \sim \text{Pois}(\lambda t)\\
\lambda|X \sim \text{Gamma}(r_0 + \text{number of observed mutations}, b_0 + \text{length of time between sequences} )
\]

Let's do prior predictive checks to see what kind of data we will observe using the above prior.

```{r}
# sample from the prior.
prior_samples_lambda = rgamma(1000,3,1)
t = number_of_months

prior_num_mutations = rpois(1000,prior_samples_lambda * t)
hist(prior_num_mutations,100,xlab = 'observed number of mutations in 21 months')
```

# Posterior distribution

Our posterior distribution is:
\[ \lambda | X \sim \text{Gamma} (3+\text{num_of_mutations}, 1 + \text{number_of_months}) 
\]

let's draw some samples from the posterior distribution and plot them and use them to calculate probabilities. For instance what is the probabililty that the mutation rate is between 1.5 and 2.5 mutations per genome per months?

```{r}
post_samples = rgamma(1000,3+num_of_mutations_egypt, 1+(12+9))
hist(post_samples,50)

sum(post_samples >1.5 & post_samples <2.5) / 1000

pgamma(2.5,3+num_of_mutations_egypt, 1+(12+9)) - pgamma(1.5,3+num_of_mutations_egypt, 1+(12+9))
```

Using the posterior samples and the analytic solution, we get very similar results. However the sampling approach is much more general.


# Extra work

How can we use the information from all 3 samples to get a better estimate about the mutation rate? For instance we could assume that these three samples are independently and identically distributed (IID). Using this assumption how would you combine the data from all 3 samples to get a better estimate for mutation rate? Maybe have a go and see if you can work this out.

Is the IID assumption a valid one?

 




