[Up to table of contents](README.md)

[Back to the previous page](How_much_of_the_genome_is_in_genes.md)

[Go the next page](Scaling_up.md)

## Where next?

From this point there are lots of things you could investigate.  For example:

- What do the 'extreme' genes look like across species?  Are large genes large in all species?
- Is there a close relationship between genome length and number of genes across species?
- Is there a relationship betwen gene length and number of exons?  Or number of transcripts?  (Within a species.)
- Which species are most gene-dense?  Which are least?
- Are there other species (like *P.falciparum*) for which most of the genome is in genes?  What are they?
- Compare particular genes of interest across species...

...and lots more questions.

To run the analysis across larger number of species, you'll need to do some work to

- tidy up the code
- make it more efficient.  (Especially in terms of disk space)

For some tips with this, see the [page on scaling up](Scaling_up.md).

To make these comparisons you will probably also want to plot some of the data. Visualisation is
beyond the scope of this tutorial (and if you're following the GMS course, will be covered later
on) but [here's a quick demonstration of how you could make a quick plot using R](Visualisation.md).

Another thing you could focus on, if you felt like it, is improving the code - for example making
it faster, slicker, better tested, or to do more comprehensive analysis. Or if you want a
programming challenge, try writing it in R instead (I'd advise using the [tidyverse
packages](https://www.tidyverse.org).

If you are on the GMS programme - we\ll ask you to pursue one or more of these ideas (or related
ideas of your own) and tell us about them in the presentation session.


